core/src/main/java/org/jruby/exceptions/RaiseException.java

```diff
@@ -91,28 +102,71 @@ public final RubyException getException() {
         return exception;
     }
 
-    private void preRaise(ThreadContext context) {
-        preRaise(context, (IRubyObject) null);
-    }
-
-    private void preRaise(ThreadContext context, IRubyObject backtrace) {
+    private void preRaise(ThreadContext context, IRubyObject backtrace, boolean capture) {
         context.runtime.incrementExceptionCount();
         if (RubyInstanceConfig.LOG_EXCEPTIONS) TraceType.logException(exception);
 
         doSetLastError(context);
         doCallEventHook(context);
 
         if (backtrace == null) {
-            backtrace = Helpers.invokeChecked(context, exception, sites(context).backtrace);
+            if (capture) { // only false to support legacy RaiseException construction (not setting trace)
+                exception.captureBacktrace(context);
+                setStackTraceFromException();
+            }
         } else {
             exception.setBacktrace(backtrace);
+            if (!backtrace.isNil() && !isEmptyArray(backtrace)) {
+                exception.captureBacktrace(context);
+            } else { // !backtrace.isNil()
+                //fillInStackTraceSkip(1 + 2); // TODO: we fill the stack due backwards compatibility
+                // we likely want a flag like -Xexception.java.trace to fine control filling in the stack-trace
+                // (but ONLY for sub-classes of RaiseException?)
+            }
```

Comment:
I think we should just remove this todo and discuss filtering separately. It would be a bonus if we can move the actual trace acquisition closer to the relevant code or convince Java not to fill in things like "fillInStackTrace".

Reply:
leftover, definitely shouldn't be in. will cleanup